### PR TITLE
feat: SDK migration single resource scraping

### PR DESCRIPTION
**Pull request by hkfgo**
This PR implements the Azure Monitor SDK migration. It migrates away from the deprecated Azure Fluent SDK to Azure SDK for .NET. Specifically, it uses the [Azure.Monitor.Query package](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/monitor/Azure.Monitor.Query) to implement integration. Since the new SDK is essentially a different wrapper around the same REST API, I'd expect identical behavior in terms of 

- Billing(free through ARM) 
- Metric querying
- Meta-level metrics to track ARM throttling 
- Meta-level metrics to track usage 

A summary of how I did it: 
- Abstracted Azure Monitor integration through the `IAzureMonitorClient` interface. 
- Implemented Azure Monitor integration using the new SDK, under the `IAzureMonitorClient` interface 
- Implemented high-level control flow to use either the new client or the legacy client, depending on feature flag 

Things I need help with: 
- If this looks like a good approach to you
- I need some pointers on how to test this :( Specifically, how do you think end-to-end testing should be performed for a major PR like this? I was thinking either integration tests or building a custom branch image and deploying to our test environment. I'm not quite sure how to do either of them though. Some help would be much appreciated!

### CODE REVIEW

File: src/Promitor.Integrations.AzureMonitor/HttpPipelinePolicies/RecordArmRateLimitMetricsPolicy.cs
Comment (hkfgo) [inline src/Promitor.Integrations.AzureMonitor/HttpPipelinePolicies/RecordArmRateLimitMetricsPolicy.cs:12]: These HttpPipelinePolicies are to replicate the legacy SDK behavior, where User-Agent header was sent before sending a request(presumably to track usage?) and ARM rate limiting metrics are recorded after response is received

File: src/Promitor.Integrations.AzureMonitor/IAzureMonitorClient.cs
Comment (hkfgo) [inline src/Promitor.Integrations.AzureMonitor/IAzureMonitorClient.cs:8]: What do you think about using this interface to toggle between the new and legacy SDK? Unfortunately both implementations of this interface are still quite similar logically. Not sure how I'd abstract more though :/

**Comment by tomkerkhove** [comment]
> I need some pointers on how to test this :( Specifically, how do you think end-to-end testing should be performed for a major PR like this? I was thinking either integration tests or building a custom branch image and deploying to our test environment. I'm not quite sure how to do either of them though. Some help would be much appreciated!

This should be covered by the testing sweet so no worries!

File: src/Promitor.Agents.Scraper/AzureMonitorClientFactory.cs
Comment (tomkerkhove) [inline src/Promitor.Agents.Scraper/AzureMonitorClientFactory.cs:33]: Nit - Let's move it as last parameter given this is temporary

File: src/Promitor.Core.Scraping/Configuration/Model/MetricsDeclaration.cs
Comment (tomkerkhove) [inline src/Promitor.Core.Scraping/Configuration/Model/MetricsDeclaration.cs:8]: Let's migrate by default.

```suggestion
        public bool UseAzureMonitorSdk { get; set; } = true;
```

File: src/Promitor.Core.Scraping/Configuration/Serialization/v1/Core/V1Deserializer.cs
Comment (tomkerkhove) [inline src/Promitor.Core.Scraping/Configuration/Serialization/v1/Core/V1Deserializer.cs:24]: Let's migrate by default, this should in theory even be technical implementation detail
```suggestion
                .WithDefault(true);
```

File: src/Promitor.Integrations.AzureMonitor/HttpPipelinePolicies/RecordArmRateLimitMetricsPolicy.cs
Comment (tomkerkhove) [inline src/Promitor.Integrations.AzureMonitor/HttpPipelinePolicies/RecordArmRateLimitMetricsPolicy.cs:58]: Move to method and re-use?

**Review by tomkerkhove — COMMENTED** [review]
Thanks for doing this, added some comments but good first step!

**Comment by hkfgo** [comment]
> > I need some pointers on how to test this :( Specifically, how do you think end-to-end testing should be performed for a major PR like this? I was thinking either integration tests or building a custom branch image and deploying to our test environment. I'm not quite sure how to do either of them though. Some help would be much appreciated!
> 
> This should be covered by the testing sweet so no worries!

Are these pipeline steps something I can repeated trigger on my own? I read the README more carefully and it seems like I can find my branch build under `:pr{pr-id}`. If I can re-trigger the CI pipeline on my own then that'd definitely cover end-to-end testing!

**Comment by tomkerkhove** [comment]
> > > I need some pointers on how to test this :( Specifically, how do you think end-to-end testing should be performed for a major PR like this? I was thinking either integration tests or building a custom branch image and deploying to our test environment. I'm not quite sure how to do either of them though. Some help would be much appreciated!
> > 
> > 
> > This should be covered by the testing sweet so no worries!
> 
> Are these pipeline steps something I can repeatedly trigger on my own? I read the README more carefully and it seems like I can find my branch build under `:pr{pr-id}`. If I can re-trigger the CI pipeline on my own then that'd definitely cover end-to-end testing!

Hey, there are docs how to run things locally here: https://github.com/tomkerkhove/promitor/blob/master/CONTRIBUTING.md#net-development

The CI also runs all these integration tests automatically in case you were wondering

**Comment by tomkerkhove** [comment]
I see you are actively working on this so I'll hold off on release!

Do let me know if you need to put this on hold

**Comment by hkfgo** [comment]
Yes I am. Getting close

File: src/Promitor.Integrations.AzureMonitor/AzureMonitorQueryClient.cs
Comment (hkfgo) [reply src/Promitor.Integrations.AzureMonitor/AzureMonitorQueryClient.cs]: I found out `AzureAuthenticationFactory` has actually dealt with authentication under `Azure.Identity` package. I switched to it and it worked :)

**Comment by hkfgo** [comment]
Good news! This most recent build appears to be working across all of our resource types. We have a good coverage within Axon, with Redis, Storage, Load Balancer, Disk, Azure SQL, Power Functions, etc. 

Let me clean up the code more(I've added so many log statements!!). Meanwhile, do you suggest anything for end-to-end testing?

**Comment by tomkerkhove** [comment]
Good news!

The automated testing already happens and is reported below, the majority of the checks are passing except for a few 👌

File: src/Promitor.Integrations.AzureMonitor/AzureMonitorQueryClient.cs
Comment (hkfgo) [reply src/Promitor.Integrations.AzureMonitor/AzureMonitorQueryClient.cs:27]: I thought about this for a bit and here's my take. While the two clients are logically almost identical, they use different SDK-specific data models. More abstraction can be achieved by defining SDK-agnostic data models and writing logical flows using those. I've done that already with configuration so it'd be doable; at the same time, I don't think it's quite worth it if the legacy client was to be removed anyways. 

Another option is, if you think it's a good idea, perhaps we can do the refactoring as a follow-up PR?

File: src/Promitor.Integrations.AzureMonitor/HttpPipelinePolicies/RegisterPromitorAgentPolicy.cs
Comment (hkfgo) [reply src/Promitor.Integrations.AzureMonitor/HttpPipelinePolicies/RegisterPromitorAgentPolicy.cs:21]: `metricSinkWriter` contains metric sink configuration which will be part of the header. It's used here
`            var promitorUserAgent = ArmUserAgent.Generate(agentVersion, _metricSinkWriter.EnabledMetricSinks);
`

File: src/Promitor.Integrations.AzureMonitor/HttpPipelinePolicies/RecordArmRateLimitMetricsPolicy.cs
Comment (hkfgo) [reply src/Promitor.Integrations.AzureMonitor/HttpPipelinePolicies/RecordArmRateLimitMetricsPolicy.cs:58]: On second thought, I don't think the synchronous path should be used at all. I decided to throw an exception there instead(and there's no more repeated code ofc), Let me know if you agree.

File: src/Promitor.Integrations.AzureMonitor/HttpPipelinePolicies/RegisterPromitorAgentPolicy.cs
Comment (hkfgo) [reply src/Promitor.Integrations.AzureMonitor/HttpPipelinePolicies/RegisterPromitorAgentPolicy.cs:30]: Same comment here, decided it makes more sense to thrown an exception on the synchronous path instead,

**Comment by hkfgo** [comment]
Looks like all tests are passing now! The failing `CodeFactor` was on the giant `MetricScraperFactory` method to find the matching scraper. There's not much we can do I think.  

Do you mind taking another look? Please ignore the log statements and modification to GitHub action for now. We can remove those before pressing the merge button

**Comment by hkfgo** [comment]
Also, any pointers on how to do remote debugging? I've found some online articles on remote debugging with VS Code + .NET + Kubernetes. Probably should have tried that to begin with instead of doing so many print statements..

File: src/Promitor.Integrations.AzureMonitor/AzureMonitorQueryClient.cs
Comment (tomkerkhove) [reply src/Promitor.Integrations.AzureMonitor/AzureMonitorQueryClient.cs:29]: Agreed, let's keep it as-is and see how it goes!

**Comment by tomkerkhove** [comment]
> Also, any pointers on how to do remote debugging? I've found some online articles on remote debugging with VS Code + .NET + Kubernetes. Probably should have tried that to begin with instead of doing so many print statements..

No, I always use VS to run the container locally and troubleshoot. If a running instance does not provide the insights you need, then we may be missing some logs

File: .github/workflows/templates-build-push-image.yml
Comment (tomkerkhove) [inline .github/workflows/templates-build-push-image.yml:1]: Let's revert this. Also, you can just use the public image that is already there (as mentioned :))

File: src/Promitor.Agents.Scraper/Scheduling/ResourcesScrapingJob.cs
Comment (tomkerkhove) [inline src/Promitor.Agents.Scraper/Scheduling/ResourcesScrapingJob.cs:285]: We are calling `_metricsDeclaration.AzureMetadata.Cloud.GetAzureEnvironment()` twice, let's move to variable and re-use that instead

File: src/Promitor.Core.Scraping/Configuration/Serialization/v1/Core/AzureMetadataDeserializer.cs
Comment (tomkerkhove) [inline src/Promitor.Core.Scraping/Configuration/Serialization/v1/Core/AzureMetadataDeserializer.cs:34]: Why do we call that specific method if we don't do anything with it?

File: src/Promitor.Core.Scraping/Configuration/Serialization/v1/Core/AzureMetadataDeserializer.cs
Comment (tomkerkhove) [inline src/Promitor.Core.Scraping/Configuration/Serialization/v1/Core/AzureMetadataDeserializer.cs:26]: Can you open an issue and link to it here please?

File: src/Promitor.Core/Metrics/MetricAggregationType.cs
Comment (tomkerkhove) [inline src/Promitor.Core/Metrics/MetricAggregationType.cs:5]: Please create a file per type, so in this case move MetricAggregationTypeConverter to a seperate file

**Review by tomkerkhove — CHANGES_REQUESTED** [review]
Requesting changes to ensure we revert the GitHub action changes, but otherwise it looks good and added some small comments.

Thanks a ton!!

File: src/Promitor.Core.Scraping/Configuration/Serialization/v1/Core/AzureMetadataDeserializer.cs
Comment (hkfgo) [reply src/Promitor.Core.Scraping/Configuration/Serialization/v1/Core/AzureMetadataDeserializer.cs:32]: The previous flow here was to validate cloud support but also transform the field type from the SDK-agnostic `AzureCloud` to `AzureEnvironment`. Here I'm just performing the validation. I'm not using the transformed `Azure.Monitor` model to keep things SDK-agnostic

File: src/Promitor.Core.Scraping/Configuration/Serialization/v1/Core/AzureMetadataDeserializer.cs
Comment (hkfgo) [reply src/Promitor.Core.Scraping/Configuration/Serialization/v1/Core/AzureMetadataDeserializer.cs]: I used  azureCloud.DetermineMetricsClientAudience() for validation, see comment above

**Comment by hkfgo** [comment]
Also, I believe there should be two quick PRs to 
1. Update documentation 
2. Make `useAzureMonitor` flag available in the Promitor chart

File: changelog/content/experimental/unreleased.md
Comment (tomkerkhove) [inline changelog/content/experimental/unreleased.md:9]: Let's reword a bit
```suggestion
- {{% tag changes %}} Migrate to latest Azure.Monitor SDK.
```

File: src/Promitor.Core/Metrics/MetricAggregationTypeConverter.cs
Comment (tomkerkhove) [inline src/Promitor.Core/Metrics/MetricAggregationTypeConverter.cs:7]: Just to confirm, did we add unit tests for these?

Comment (hkfgo) [reply src/Promitor.Core/Metrics/MetricAggregationTypeConverter.cs:7]: Not previously but I added them just now :)

**Comment by hkfgo** [comment]
Promitor documentation PR: https://github.com/promitor/docs/pull/62
Promitor chart PR: https://github.com/promitor/charts/pull/168

I believe they are dependencies of the next release but not this PR getting merged. I'm making this distinction because I'm waiting on merge to master to rebase and continue batch scraping work. No rush though. Thanks!

**Review by tomkerkhove — APPROVED** [review]
Thanks a ton for doing this and not giving up after all my comments!

**Pull request merged**
